wen-restart/src/heaviest_fork_aggregate.rs

```diff
@@ -0,0 +1,555 @@
+pub struct HeaviestForkFinalResult {
+    pub block_stake_map: HashMap<(Slot, Hash), u64>,
+    pub total_active_stake: u64,
+    pub total_active_stake_seeing_supermajority: u64,
+}
+
+impl HeaviestForkAggregate {
+    pub(crate) fn new(
+        wait_for_supermajority_threshold_percent: u64,
+        my_shred_version: u16,
+        epoch_stakes: &EpochStakes,
+        heaviest_fork_slot: Slot,
+        heaviest_fork_hash: Hash,
+        my_pubkey: &Pubkey,
+    ) -> Self {
+        let mut active_peers = HashSet::new();
+        active_peers.insert(*my_pubkey);
+        let mut block_stake_map = HashMap::new();
+        block_stake_map.insert(
+            (heaviest_fork_slot, heaviest_fork_hash),
+            Self::validator_stake(epoch_stakes, my_pubkey),
+        );
+        Self {
+            supermajority_threshold: wait_for_supermajority_threshold_percent as f64 / 100.0,
+            my_shred_version,
+            my_pubkey: *my_pubkey,
+            epoch_stakes: epoch_stakes.clone(),
+            heaviest_forks: HashMap::new(),
+            block_stake_map,
+            active_peers,
+            active_peers_seeing_supermajority: HashSet::new(),
+        }
+    }
+
+    fn validator_stake(epoch_stakes: &EpochStakes, pubkey: &Pubkey) -> u64 {
+        epoch_stakes
+            .node_id_to_vote_accounts()
+            .get(pubkey)
+            .map(|x| x.total_stake)
+            .unwrap_or_default()
+    }
+
+    pub(crate) fn aggregate_from_record(
+        &mut self,
+        key_string: &str,
+        record: &HeaviestForkRecord,
+    ) -> Result<Option<HeaviestForkRecord>> {
+        let from = Pubkey::from_str(key_string)?;
+        let bankhash = Hash::from_str(&record.bankhash)?;
+        let converted_record = RestartHeaviestFork {
+            from,
+            wallclock: record.wallclock,
+            last_slot: record.slot,
+            last_slot_hash: bankhash,
+            observed_stake: record.total_active_stake,
+            shred_version: record.shred_version as u16,
+        };
+        Ok(self.aggregate(converted_record))
+    }
+
+    pub(crate) fn aggregate(
+        &mut self,
+        new_heaviest_fork: RestartHeaviestFork,
+    ) -> Option<HeaviestForkRecord> {
+        let total_stake = self.epoch_stakes.total_stake();
+        let from = &new_heaviest_fork.from;
+        let sender_stake = Self::validator_stake(&self.epoch_stakes, from);
+        if sender_stake == 0 {
+            warn!(
+                "Gossip should not accept zero-stake RestartLastVotedFork from {:?}",
+                from
+            );
+            return None;
+        }
+        if new_heaviest_fork.shred_version != self.my_shred_version {
+            warn!(
+                "Gossip should not accept RestartLastVotedFork with different shred version {} from {:?}",
+                new_heaviest_fork.shred_version, from
+            );
+            return None;
+        }
+        self.active_peers.insert(*from);
```

Comment:
Moved update of active_peers to later in the function. Right now all operations of heaviest_fork_aggregate.rs are done in a single thread loop in wen_restart, same applies to last_voted_fork_slots_aggregate.rs. That's why we didn't need any locks in here. Otherwise I think moving them into a separate function is not enough, we also need to add locks around them.

Reply:
> That's why we didn't need any locks in here. Otherwise I think moving them into a separate function is not enough, we also need to add locks around them.

Oh I was less concerned about locking/other threads modifying it, and more concerned about code maintainability.

I could easily imagine we might introduce a change that causes an early exit after the `active_peers.insert` and  before we update the `block_stake_map`, which would lead to an inconsistency